cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceService.java

```diff
@@ -0,0 +1,215 @@
+   */
+  private ListenableFuture<Void> asyncCreateDefaultNamespace() {
+    ListeningExecutorService executorService = MoreExecutors.listeningDecorator(
+      Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("default-namespace-ensurer")));
+    return executorService.submit(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        createDefaultNamespace();
+        return null;
+      }
+    });
+  }
+  
+  /**
+   * Creates the default namespace at a more deterministic time.
+   * It should be removed once we stop support for v2 APIs, since 'default' namespace is only reserved for v2 APIs.
+   */
+  private void createDefaultNamespace() {
+    NamespaceMeta.Builder builder = new NamespaceMeta.Builder();
+    NamespaceMeta defaultNamespace = builder.setId(Constants.DEFAULT_NAMESPACE)
+      .setName(Constants.DEFAULT_NAMESPACE)
+      .setDescription(Constants.DEFAULT_NAMESPACE)
+      .build();
+
+    try {
+      createNamespace(defaultNamespace);
+      LOG.info("Successfully created 'default' namespace.");
+    } catch (AlreadyExistsException e) {
+      LOG.info("'default' namespace already exists.");
+    }
+  }
+
+  /**
+   * Lists all namespaces
+   *
+   * @return a list of {@link NamespaceMeta} for all namespaces
+   */
+  public List<NamespaceMeta> listNamespaces() {
+    startAndWait();
+    return store.listNamespaces();
+  }
+
+  /**
+   * Gets details of a namespace
+   *
+   * @param namespaceId {@link Id.Namespace} of the requested namespace
+   * @return the {@link NamespaceMeta} of the requested namespace
+   * @throws NotFoundException if the requested namespace is not found
+   */
+  public NamespaceMeta getNamespace(Id.Namespace namespaceId) throws NotFoundException {
+    startAndWait();
+    NamespaceMeta ns = store.getNamespace(namespaceId);
+    if (ns == null) {
+      throw new NotFoundException(NAMESPACE_ELEMENT_TYPE, namespaceId.getId());
+    }
+    return ns;
+  }
+
+  /**
+   * Creates a new namespace
+   *
+   * @param metadata the {@link NamespaceMeta} for the new namespace to be created
+   * @throws AlreadyExistsException if the specified namespace already exists
+   */
+  public void createNamespace(NamespaceMeta metadata) throws AlreadyExistsException {
+    startAndWait();
+    NamespaceMeta existing = store.createNamespace(metadata);
+    if (existing != null) {
+      throw new AlreadyExistsException(NAMESPACE_ELEMENT_TYPE, metadata.getId());
+    }
+  }
+
+  /**
+   * Deletes the specified namespace
+   *
+   * @param namespaceId the {@link Id.Namespace} of the specified namespace
+   * @throws NotFoundException if the specified namespace does not exist
+   */
+  public void deleteNamespace(Id.Namespace namespaceId) throws NotFoundException {
+    startAndWait();
+    // Delete Preferences associated with this namespace
+    preferencesStore.deleteProperties(namespaceId.getId());
```

Comment:
Thanks for catching that. I think we need a bigger change here. Deleting preferences, dashboard settings and namespace will have to be done in a single transaction, preferably in ``DefaultStore`` or in a separate ``NamespaceStore``. Is it ok if I take care of it in a separate PR - as part of https://issues.cask.co/browse/CDAP-870, since currently this code is just moved over from ``NamespaceHttpHandler``? We need to do a lot more in namespace delete, like deleting apps, datasets, etc.